### ISSUE TITLE
Questions regarding data shuffling in the preprocessing step

### ISSUE THREAD

**Issue by PersianNLPer** [OpenNMT/OpenNMT-py#371]
Hi,
I have two questions regarding the data preprocessing, in particular the data shuffling part.
1. It seems that even though there is a switch in the preprocess_opts() method in opts.py for shuffling the data (-shuffle), it is never used in the preprocessing step. Am I right?
2. Since this switch should be a boolean (storing only true or false), I wonder why it is defined such that it takes integers with the default value of 1. At first I thought that this value might be used as a seed for the random generator which shuffles the data, but this does not make sense cause if we always shuffle the data with the same seed, the order will be always the same, hence the system does not see different combinations of the training data.
Any idea about these questions, in particular the first one which can greatly affect the training procedure?

Thanks,
Amin

**Comment by JianyuZhan** [comment]
1. Right.  Shuffling has been done in `torchtext.data.iterator`, when we iterate over the examples.
2. Sorry, I've tried to dig into the git history but found no answer.

**Comment by PersianNLPer** [comment]
Thank you JianyuZhan.
I agree that `torchtext.data.iterator` should take care of shuffling, but as I see in the `__init__` function of `class Iterator(object):` the default value for `shuffle` is `None`. And I can not see the part of the code that initializes this argument. So, I am not sure how this argument gets the proper value to shuffle the data. 
May be I am missing something.
So, can you please point me to the part which sets the shuffle flag to True?

Thanks,
Amin

**Comment by JianyuZhan** [comment]
in `torchtext/data/iterator.py`:86
```
self.shuffle = train if shuffle is None else shuffle
```

`self.shuffle`  would be set to True, if we are in train mode.

**Comment by PersianNLPer** [comment]
Ok. So, by default we apply the shuffling during the training. Right?

**Comment by JianyuZhan** [comment]
Right, I think this is why the `shuffle` option is dangling. As for why it is an int, it is an archaeology problem, on which I think it doesn't make sense to put too much effort ~

**Comment by PersianNLPer** [comment]
Thank you JianyuZhan!
So, I am going to close the issue.
